vulnerabilities/api.py

```diff
@@ -193,7 +224,27 @@ class Meta:
 
 
 class VulnerabilityViewSet(viewsets.ReadOnlyModelViewSet):
-    queryset = Vulnerability.objects.all()
+    def get_queryset(self):
+        params = self.request.query_params
+        query = Q()
+        name = params.get("name")
+        if name:
+            query &= Q(name=name)
+        namespace = params.get("namespace")
+        if namespace:
+            query &= Q(namespace=namespace)
+        type = params.get("type")
+        if type:
+            query &= Q(type=type)
+        queryset = Vulnerability.objects.prefetch_related(
+            Prefetch(
+                "packages",
+                queryset=Package.objects.filter(query, packagerelatedvulnerability__fix=True),
+                to_attr="filtered_fixed_packages",
+            )
+        )
+        return queryset
```

Comment:
This is still duplicated code and much harder to read and understand than the previous code.
This is not what I meant by "proper filters" but I was rather referring to https://www.django-rest-framework.org/api-guide/filtering/#djangofilterbackend.
Any reason to not use a FilterSet here?

Reply:
The filters applied by me are for Package model and if the viewset is on Vulnerabilities model

Reply:
I understand, but relational fields are not doable on FilterSet?

Reply:
I can do something like this https://github.com/philipn/django-rest-framework-filters#filtering-across-relationships, but the filter will then look like `api?package__name=name` not `api?name=name`

Reply:
> `api?package__name=name`

Is it a problem?

Reply:
@TG1999 you already defined a PackageFilterSet, you should use it, for example:

```
class VulnerabilityViewSet(viewsets.ReadOnlyModelViewSet):
    def get_fixed_packages_qs(self):
        package_filter_data = {"packagerelatedvulnerability__fix": True}

        query_params = self.request.query_params
        for field_name in ["name", "namespace", "type"]:
            value = query_params.get(field_name)
            if value:
                package_filter_data[field_name] = value

        return PackageFilterSet(package_filter_data).qs

    def get_queryset(self):
        queryset = Vulnerability.objects.prefetch_related(
            Prefetch(
                "packages",
                queryset=self.get_fixed_packages_qs(),
                to_attr="filtered_fixed_packages",
            )
        )
        return queryset
```

Also, make sure to add `packagerelatedvulnerability__fix` to the PackageFilterSet.fields.

Lastly, you should always document such methods using the docstring, it is not obvious to the reader what is happening here and why we are prefetching filtered package data.